File: src/commands/edit.ts
import { spawn } from 'node:child_process';
import { parseArgs } from 'node:util';

import { Ticket } from '../lib/ticket.ts';
import { exitWithError } from './utils.ts';

export async function cmdEdit(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 1) {
    exitWithError('Usage: tk edit <id>');
  }

  const ticket = await Ticket.find(positionals[0]);

  if (process.stdin.isTTY) {
    const editor = process.env.EDITOR || 'vi';
    await new Promise<void>((resolve, reject) => {
      const proc = spawn(editor, [ticket.filePath], { stdio: 'inherit' });
      proc.on('close', (code) => {
        code === 0 ? resolve() : reject(new Error(`Editor exited with code ${code}`));
      });
    });
  } else {
    console.log(`Edit ticket file: ${ticket.filePath}`);
  }
}

File: src/commands/edit.rs
use serde_json::json;

use super::{CommandOutput, open_in_editor_for_entity};
use crate::error::Result;
use crate::ticket::Ticket;

/// Open a ticket in the default editor
pub async fn cmd_edit(id: &str, output_json: bool) -> Result<()> {
    let ticket = Ticket::find(id).await?;

    // Output in JSON format if requested (skip editor)
    if output_json {
        return CommandOutput::new(json!({
            "id": ticket.id,
            "file_path": ticket.file_path.to_string_lossy(),
            "action": "edit",
        }))
        .print(output_json);
    }

    open_in_editor_for_entity("ticket", &ticket.file_path, output_json)
}
